### DISCUSSION TITLE
MLA - Partial statistics

### DISCUSSION THREAD

**Discussion by clickersmudge** [kubermatic/kubermatic#12126] [Q&A]
Hi,
We installed the MLA stack on the Master/ Seed cluster, and MLA for the user clusters.

# Master / Seed
1. Securing System Services
2. IAP

3. Grafana
https://grafana.kubermatic.example.com
![Zrzut ekranu z 2023-04-04 10-10-41](https://user-images.githubusercontent.com/113194014/229729608-f8fab9bf-2394-4d34-8ca7-0328b62de1ff.png)
Working

# User cluster

1. Enabled MLA
```
kubectl edit kubermaticconfiguration --namespace kubermatic kubermatic
  spec.featureGates.UserClusterMLA: true
nano seed.yaml
  spec.mla.userClusterMLAEnabled: true
kubectl apply --filename seed.yaml
```
![obraz](https://user-images.githubusercontent.com/113194014/229730314-c544f2d5-1c1a-4c84-ac4c-05eaa06c01f7.png)

![obraz](https://user-images.githubusercontent.com/113194014/229730479-e3cfccf6-7d70-41e3-b4ef-d2f6e63a1677.png)

![obraz](https://user-images.githubusercontent.com/113194014/229730571-9a5f52e3-b199-4007-b3f9-bddea8fc56e7.png)

2. We set up external repositories on Wasabi S3.
![obraz](https://user-images.githubusercontent.com/113194014/229731311-4677098c-9e8d-4aa2-a21c-9dcb1ef77905.png)

3. Installation
```
./kubermatic-installer deploy usercluster-mla --config kubermatic.yaml --helm-values values.mla.yaml \
  --mla-force-secrets \
  --mla-skip-minio \
  --mla-skip-minio-lifecycle-mgr
```
By the way, Minio, despite the "--mla-skip-minio" flag, installs anyway.

4. Grafana
https://grafana.kubermatic.kubermatic.example.com
General / Kubernetes Overview Dashboard
![obraz](https://user-images.githubusercontent.com/113194014/229733566-d215d274-90be-4f6a-be3a-7df2913341cc.png)

# Problem
Other dashboards do not work.

![obraz](https://user-images.githubusercontent.com/113194014/229732298-8646bc23-d3a7-4d91-ba3a-ac7483aa0f22.png)

![obraz](https://user-images.githubusercontent.com/113194014/229732542-0c54d57e-40d6-4242-963e-af65bfb461a4.png)

![obraz](https://user-images.githubusercontent.com/113194014/229732669-2e107c20-cd32-48ff-8ece-6b094d000b5a.png)

All pods in namespace "mla" work.
Not sure where to debug?

**Comment by wurbanski** [comment]
Hi,

> By the way, Minio, despite the "--mla-skip-minio" flag, installs anyway.

Thank you, we will verify this and fix the problem.

> All pods in namespace "mla" work.
Not sure where to debug?

Have you tried scheduling some more pods on the cluster? system pods should be excluded from the dashboards by design.

**Reply by clickersmudge** [reply]
Hi Wojtek we will check this. / Cześć Wojtek, sprawdzimy.

The strange thing is that the default dashboard displays "node and namespace" correctly. 
![Zrzut ekranu z 2023-04-04 12-25-16](https://user-images.githubusercontent.com/113194014/229765079-1457672d-9d52-4bd2-b9e8-0fa2e83ca3c5.png)
![Zrzut ekranu z 2023-04-04 12-25-24](https://user-images.githubusercontent.com/113194014/229765119-c6593165-4aca-43d5-90b5-1bba72d8d07e.png)

But "General / Nodes Overview" not working.
![Zrzut ekranu z 2023-04-04 12-26-03](https://user-images.githubusercontent.com/113194014/229765341-737d4be7-8dcc-4747-b7b0-363325354a03.png)

**Reply by wurbanski** [reply]
>  Cześć Wojtek, sprawdzimy.

👋 :)

>But "General / Nodes Overview" not working.

Do you have `node-exporter` add-on enabled for your user cluster? It is required for for the MLA dashboards to work correctly.

**Reply by clickersmudge** [reply]
Yes. We added node-exporter and kube-state-metrics to the clusters.
![Zrzut ekranu z 2023-04-04 13-06-07](https://user-images.githubusercontent.com/113194014/229773427-4d9139cb-ef73-475f-a36f-074fe9c1fe09.png)
![Zrzut ekranu z 2023-04-04 13-04-58](https://user-images.githubusercontent.com/113194014/229773471-0afe9424-e320-42de-826e-4174895b1bb0.png)
![Zrzut ekranu z 2023-04-04 13-05-07](https://user-images.githubusercontent.com/113194014/229773479-ab8d7c29-b5ab-42e1-9bd4-99df7bb16704.png)

**Comment by clickersmudge** [comment]
Hi,
We have been debugging for several days and cannot understand the cause of the error.

# Dubugging scenario
## Uninstalling old MLA
```
helm list --namespace mla
helm uninstall --namespace mla alertmanager-proxy consul cortex grafana loki-distributed minio minio-lifecycle-mgr mla-secrets
kubectl delete namespace mla
```

##  values.mla.yaml

Default values

```
mlaSecrets:
  minio:
    ## Set enabled to false if you want to re-use an existing secret for minio.
    enabled: true
    ## Set default accesskey, secretkey
    ## AccessKey and secretKey is generated when not set
    accessKey: "accesskey"
    secretKey: "secretkey"
  grafana:
    ## Set enabled to false if you want to re-use an existing secret for grafana.
    enabled: true
    adminUser: admin
    adminPassword: strongpassword
```

## Install

With default Minio. Without external S3.

```
./kubermatic-installer deploy usercluster-mla --config kubermatic.yaml --helm-values values.mla.yaml \
  --mla-force-secrets
```

## Results

![obraz](https://user-images.githubusercontent.com/113194014/230606642-8b20bb86-1be2-4aa2-a388-88a37fb8acce.png)
![obraz](https://user-images.githubusercontent.com/113194014/230606679-a5e395e2-68ad-4e35-b4cb-155eadb27d65.png)

## Problem
Dashboard is empty.
![obraz](https://user-images.githubusercontent.com/113194014/230606752-f2b0d072-7815-4d70-a778-edbc7d5539b9.png)

Cannot switching between Grafana Organizations as in the documentation (https://docs.kubermatic.com/kubermatic/v2.22/tutorials-howtos/monitoring-logging-alerting/user-cluster/user-guide/). Switching button missing.
https://docs.kubermatic.com/kubermatic/v2.22/tutorials-howtos/monitoring-logging-alerting/user-cluster/user-guide/
Log in via the KKP button.

**Comment by clickersmudge** [comment]
Hi,
@wurbanski 
I wanted to ask if you are sure the dashboards in the "charts/mla/grafana/files" folder are working out of the box.

# Dashboards debugging
kkp-kubernetes-nodes.json (not working)
kkp-kubernetes-nodes-overview.json (not working)
kkp-kubernetes-pods.json (not working)
kkp-kubernetes-statefulsets.json (not working)
kubernetes-overview.json (working, but not all)

# kubernetes-overview.json (working, but not all)
Once installed
![obraz](https://user-images.githubusercontent.com/113194014/231255868-74fecac9-7ca4-4cda-be57-e2abf266d231.png)
Pods CPU usage does not work.

We did a bit of searching on the web and in reference to this https://github.com/prometheus-operator/kube-prometheus/issues/1695 topic we have changed:
```
sum (rate (container_cpu_usage_seconds_total{image!="",name=~"^k8s_.*",kubernetes_io_hostname=~"^$Node$",namespace=~"^$namespace$"}[1m])) by (pod)
```
to
```
sum (rate (container_cpu_usage_seconds_total{image!="",kubernetes_io_hostname=~"^$Node$",namespace=~"^$namespace$"}[1m])) by (pod)
```
And now it works.
![obraz](https://user-images.githubusercontent.com/113194014/231256591-961f23a2-ade1-468e-9440-7520f364309e.png)


# kkp-kubernetes-pods.json (not working)
We noticed that the working dashboard (kubernetes-overview.json) has in the variables.
![obraz](https://user-images.githubusercontent.com/113194014/231256950-54706f55-34b0-43e2-9e25-30f2bbf7e4c9.png)
```
label_values(namespace)
```

Other e.g. kkp-kubernetes-pods.json
![obraz](https://user-images.githubusercontent.com/113194014/231257268-3273abdc-9e3e-4ae5-9a6e-0ab6afeaef44.png)
```
label_values(kube_pod_info, namespace)
```
This query does not work.

**Comment by clickersmudge** [comment]
# Solution

We uninstalled node-exporter and kube-state-metrics installed by KKP.
![obraz](https://user-images.githubusercontent.com/113194014/231582459-bf35cbc5-f78e-4cdb-b613-096f414549ae.png)

We installed charts on the test cluster from the "charts/monitoring/node-exporter" and "charts/monitoring/kube-state-metrics"folder.
It seems to work, and those installed by KKP cause some problems. We'll final test tomorrow....
![obraz](https://user-images.githubusercontent.com/113194014/231582169-9aa3546e-0d4d-41a1-97a4-eb66e9cb9aa8.png)

**Reply by wurbanski** [reply]
Hi,

Sorry for responding late. Thanks for your input, this solution already brings a lot of information as to what exactly is wrong - I'll verify it on our end as well and get back with more info.

**Comment by clickersmudge** [comment]
Hi,
Thanks for your interest.
We discovered that the problem is with 'kube-state-metrics' and 'node-exporter'. Those installed from the panel do not work.

We manually installed the version from the KKP folder "charts/monitoring/kube-state-metrics/" and "charts/monitoring/node-exporter/". The stats from "/kube-state-metrics" are working. Those from "node-exporter" do not. We think they are installing too old versions of charts.
